components/canvas/vello_cpu_backend.rs

```diff
@@ -0,0 +1,452 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+use std::cell::RefCell;
+use std::collections::HashMap;
+use std::sync::Arc;
+
+use canvas_traits::canvas::{
+    CompositionOptions, FillOrStrokeStyle, FillRule, LineOptions, Path, ShadowOptions,
+};
+use compositing_traits::SerializableImageData;
+use euclid::default::{Point2D, Rect, Size2D, Transform2D};
+use fonts::{ByteIndex, FontIdentifier, FontTemplateRefMethods as _};
+use ipc_channel::ipc::IpcSharedMemory;
+use kurbo::Shape;
+use pixels::{Snapshot, SnapshotAlphaMode, SnapshotPixelFormat};
+use range::Range;
+use vello_cpu::{kurbo, peniko};
+use webrender_api::{ImageDescriptor, ImageDescriptorFlags};
+
+use crate::backend::{Convert, GenericDrawTarget};
+use crate::canvas_data::{Filter, TextRun};
+
+thread_local! {
+    /// The shared font cache used by all canvases that render on a thread. It would be nicer
+    /// to have a global cache, but it looks like font-kit uses a per-thread FreeType, so
+    /// in order to ensure that fonts are particular to a thread we have to make our own
+    /// cache thread local as well.
+    static SHARED_FONT_CACHE: RefCell<HashMap<FontIdentifier, peniko::Font>> = RefCell::default();
+}
+
+pub(crate) struct VelloCPUDrawTarget {
+    /// Because this is stateful context
+    /// caller cannot assume anything about transform, paint, stroke,
+    /// so it should provide it's own used by each command
+    /// but it can assume paint_transform to be identity
+    /// and fill rule to be `peniko::Fill::NonZero`
+    ///
+    /// This is because paint_transform is rarely set,
+    /// so it's cheaper to always reset it after use.
+    ctx: vello_cpu::RenderContext,
+    pixmap: vello_cpu::Pixmap,
+    clips: Vec<Path>,
+}
+
+impl VelloCPUDrawTarget {
+    fn with_composition(
+        &mut self,
+        composition_options: &CompositionOptions,
+        f: impl FnOnce(&mut Self),
+    ) {
+        self.ctx.push_layer(
+            None,
+            Some(composition_options.composition_operation.convert()),
+            Some(composition_options.alpha as f32),
+            None,
+        );
+        f(self);
+        self.ctx.pop_layer();
+    }
+
+    fn ignore_clips(&mut self, f: impl FnOnce(&mut Self)) {
+        // pop all clip layers
+        for _ in &self.clips {
+            self.ctx.pop_layer();
+        }
+        f(self);
+        // push all clip layers back
+        for path in &self.clips {
+            self.ctx.push_clip_layer(&path.0);
+        }
+    }
+
+    fn pixmap(&mut self) -> &[u8] {
+        //self.scene.flush();
```

Comment:
```suggestion
```

Reply:
I left this because we will need it, if we ever use multithreaded rendering. But I remembered that in case of single threaded rendering flush is nop, so we should just always call it. Fixed in https://github.com/servo/servo/compare/a5ce03131fb7ee34aee8bf5eec6c63d1c5dee5a6..41ad087bc0ce56bfcff6a38b59b586ecaf19c839